Disallow {FakeTensor,FunctionalTensor}.data_ptr

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122514

This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @ezyang @gchanan @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

Differential Revision: [D55366199](https://our.internmc.facebook.com/intern/diff/D55366199)

zou3519 committed:
Disallow {FakeTensor,FunctionalTensor}.data_ptr

This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

ezyang commented on c10/core/StorageImpl.h:
```diff
@@ -118,12 +119,12 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
-    maybe_materialize_cow();
-    return data_ptr_;
-  }
+  const at::DataPtr& data_ptr() const;
```
mehhh, you sure you don't want this inlining? What would be more normal is to do an inlinable conditional, where the slow path goes into a non-inlined function

zou3519 replied:
I didn't realize that putting things in a header automatically inlined them without the inline keyword. I moved this around so that I could develop faster without a bunch of recompiles from including this header everywhere. I'll put them back

zou3519 replied:
#til https://en.cppreference.com/w/cpp/language/inline#:~:text=A%20function%20defined%20entirely%20inside,(since%20C%2B%2B20).

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang ezyang gchanan voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

ezyang commented on torch/_dynamo/trace_rules.py:
```diff
@@ -1183,7 +1185,6 @@
         "torch._C._test_only_populate_upgraders",
         "torch._C._test_only_remove_entry_to_op_version_map",
         "torch._C._test_only_remove_upgraders",
-        "torch._C._to_dlpack",
```
oh?

zou3519 replied:
Previously, this was allow_in_graph'ed, but would always access FakeTensor.data_ptr and then graph break due to the return being a Dynamo-unsupported type. This PR makes it so that we alays graph break on `torch._C._to_dlpack`. The end result is the same (graph break) but we end up avoiding the FakeTensor.data_ptr access.

ezyang commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -127,7 +127,19 @@ def _register_tensor_wrapper(tensor) -> None:
+    ):
+        data_ptr = 0
+    else:
+        data_ptr = tensor.elem.data_ptr()
```
cc @yifuwang

bdhirsh replied:
Yeah this data_ptr -> work map getting filled in when we trace through `AsyncCollectiveTensor` during PT2 tracing is almost definitely one of the reasons `AsyncCollectiveTensor` (supposedly) doesn't work great with `torch.compile` today at the graph boundaries. I told @wanchaol I would try to take a look at this one some other DTensor work wraps up, although probably not in the next month or two

albanD replied:
Here as well, should this be a WeakStorageIdDict ?

yifuwang replied:
FYI this code has been deprecated and I'm planning to remove it soon.

The tensor -> work tracking has been moved to C++ and the registry uses `getWeakStorageImpl()` as the key.

https://github.com/pytorch/pytorch/blob/e61aaab72562fe95d76e77b01678d71467e1739e/torch/csrc/distributed/c10d/Functional.cpp#L15-L24

ezyang commented on test/torch_np/numpy_tests/core/test_dlpack.py:
```diff
@@ -46,7 +48,8 @@ def test_dunder_dlpack_refcount(self):
 
-    @xpassIfTorchDynamo  # (reason="pytorch does not raise")
+    @unittest.expectedFailure
+    @skipIfTorchDynamo("I can't figure out how to get __dlpack__ into trace_rules.py")
```
cc @yanboliang

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55366199).

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -819,8 +821,10 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        allow_unsafe_data_ptr_access=True,
```
Just checking: if this flag exists basically just to only ban data ptr access in PT2, but the long term plan is to unconditionally ban data ptr access on FakeTensor, do we want to mark this argument private so we can more easily kill it later?

zou3519 replied:
hmm

zou3519 replied:
Yeah that sounds reasonable

bdhirsh commented on torch/autograd/graph.py:
```diff
@@ -536,11 +536,33 @@ def wrapped_fn(grad: torch.Tensor):
 
 def _get_tid(t) -> Tuple[int, int, int]:
-    return (id(t), t.data_ptr(), t._version)
+    # FIXME: This is almost definitely a bug.
```
😮 does the autograd engine use data_ptr somewhere

soulitzer replied:
Hmm not that I know of. For context though, the code here is a utility independent of the autograd engine that uses TorchDispatchMode to dynamically insert clones whenever user does in-place on a tensor that has been saved for backward. We're using data_ptr as a way to track storage aliasing, so if a particular tensor is in-placed, any tensor saved for backward aliasing that tensor can also be cloned.

albanD replied:
We have storage pyobject preservation now @ezyang told me.
We can use a that as a unique id instead of data_ptr!

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang ezyang gchanan voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

Differential Revision: [D55366199](https://our.internmc.facebook.com/intern/diff/D55366199)

[ghstack-poisoned]

zou3519 committed:
Update on "Disallow {FakeTensor,FunctionalTensor}.data_ptr"


This PR:
- disallows FakeTensor.data_ptr when it is called inside PT2 or fx tracing.
- disallows FunctionalTensor.data_ptr (python FunctionalTensor is only used in
  PT2)

The motivation behind this is that the leading cause of segfaults when
using custom ops with PT2 is calling .data_ptr on FunctionalTensor or
FakeTensor.

This change is BC-breaking. If your code broke as a result of this, it's
because there was a bug in it (these .data_ptr should never be
accessed!). You can either fix the bug (recommended) or get the previous
behavior back with:
```
from torch._subclasses.fake_tensor import FakeTensor
from torch._subclasses.functional_tensor import FunctionalTensor

data_ptr = 0 if isinstance(tensor, (FakeTensor, FunctionalTensor)) else tensor.data_ptr()
```

Test Plan:
- existing tests

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang ezyang gchanan voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

Differential Revision: [D55366199](https://our.internmc.facebook.com/intern/diff/D55366199)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55366199).

albanD commented on c10/core/StorageImpl.cpp:
```diff
@@ -11,6 +11,16 @@ C10_API std::array<StorageImplCreateHelper, at::COMPILE_TIME_MAX_DEVICE_TYPES>
+void throwNullDataPtrError() {
+  TORCH_CHECK(
+      false,
+      "Cannot access data pointer of Tensor (e.g. FakeTensor, FunctionalTensor). "
```
nit: use exact method name `data_ptr()`

albanD commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -470,6 +470,8 @@ def __new__(cls, fake_mode, elem, device, constant=None):
             device_for_backend_keys=device,
         )
+        if not fake_mode._allow_unsafe_data_ptr_access:
+            torch._C._set_throw_on_mutable_data_ptr(self)
```
Why only on mutable data_ptr and not all access?

zou3519 replied:
It's mostly that I'm minimizing the blast radius

albanD commented on torch/_tensor.py:
```diff
@@ -378,9 +378,18 @@ def _reduce_ex_internal(self, proto):
             return (torch._utils._rebuild_nested_tensor, args_nested)
         elif (
-            self.data_ptr() == 0
-            and type(self) is not torch.Tensor
```
please just re-order this check and the one above isntead of special casing!

zou3519 replied:
Was reordered

albanD reviewed (approved):
Only small things and fix to particular modules can be done in follow ups.

As discussed, we should have a plan to enable this both in PT2 and eager!

yifuwang reviewed (approved):
LG from functional collective's perspective.

kurtamohler reviewed (approved):
Changes to COW code look good to me

zou3519 commented:
@pytorchbot merge
